pkg/cloud/services/secretsmanager/secret.go

```diff
@@ -0,0 +1,114 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secretsmanager
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	apirand "k8s.io/apimachinery/pkg/util/rand"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/awserrors"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/converters"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+)
+
+const (
+	parameterPrefix  = "aws.cluster.x-k8s.io"
+	namespacesPrefix = "namespaces"
+	clustersPrefix   = "clusters"
+	instancesPrefix  = "instances"
+)
+
+// Create stores a secret in AWS Secrets Manager for a given machine
+func (s *Service) Create(m *scope.MachineScope, data []byte) (*string, error) {
+	// Make sure to use the MachineScope here to get the merger of AWSCluster and AWSMachine tags
+	additionalTags := m.AdditionalTags()
+	// Set the cloud provider tag
+	additionalTags[infrav1.ClusterAWSCloudProviderTagKey(s.scope.Name())] = string(infrav1.ResourceLifecycleOwned)
+
+	tags := infrav1.Build(infrav1.BuildParams{
+		ClusterName: s.scope.Name(),
+		Lifecycle:   infrav1.ResourceLifecycleOwned,
+		Name:        aws.String(m.Name()),
+		Role:        aws.String(m.Role()),
+		Additional:  additionalTags,
+	})
+
+	name, err := s.secretName(m)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.scope.SecretsManager.CreateSecret(&secretsmanager.CreateSecretInput{
+		Name:         aws.String(name),
+		SecretBinary: data,
+		Tags:         converters.MapToSecretsManagerTags(tags),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.ARN, nil
+}
+
+// Delete the secret belonging to a machine from AWS Secrets Manager
+func (s *Service) Delete(m *scope.MachineScope) error {
+	secretArn := m.AWSMachine.Spec.CloudInit.SecretARN
+	if secretArn == nil {
+		return errors.New("no secret for machine")
+	}
+	_, err := s.scope.SecretsManager.DeleteSecret(&secretsmanager.DeleteSecretInput{
+		SecretId:                   m.AWSMachine.Spec.CloudInit.SecretARN,
+		ForceDeleteWithoutRecovery: aws.Bool(true),
+	})
+
+	if awserrors.IsNotFound(err) {
+		return nil
+	}
+
+	return err
+}
+
+func (s *Service) secretName(m *scope.MachineScope) (string, error) {
+	prefix := strings.Join(
+		[]string{
+			parameterPrefix,
+			namespacesPrefix,
+			s.scope.Namespace(),
+			clustersPrefix,
+			s.scope.Name(),
+			instancesPrefix,
+			m.Name(),
+		},
+		"/",
+	) + "-"
+
+	randomLen := 512 - len(prefix)
+
+	// apirand uses 27 runes, 27^27 is closest to 2^128 for 128-bits of entropy.
+	if randomLen < 27 {
+		return "", errors.New("insufficient entropy for AWS secret name")
```

Comment:
In practice, would this ever happen? This isn't something the user has control over. What can we do to ensure we never hit this? What is the impact of using a hard-coded length instead of a dynamic randomLen?

Reply:
Switched to a static 54 chars for 256-bits.